### PR TITLE
ARROW-11665: [C++][Python] Improve docstrings for decimal and union types

### PR DESCRIPTION
**Pull request by pitrou**
Also add Python factory functions `pa.sparse_union` and `pa.dense_union`.

### CODE REVIEW

**Comment by pitrou** [comment]
@jonkeane Can you take a look at this?

File: cpp/src/arrow/type.h [hunk 1]
```diff
@@ -915,6 +926,16 @@ class ARROW_EXPORT Decimal128Type : public DecimalType {
...
+///
+/// Arrow decimals are fixed-point decimal numbers encoded as a scaled
+/// integer.  The precision is the number of significant digits that the
+/// decimal type can represent; the scale is the number of digits after
+/// the decimal point (note the scale can be negative).
+///
+/// For most use cases, the maximum precision offered by Decimal128Type
+/// is sufficient, and it will result in a more compact and more efficient
+/// encoding.  Decimal256Type is useful if you need a precision higher
+/// than 38 significant digits.
```
Comment (jonkeane) [inline cpp/src/arrow/type.h:938]: Would it be possible to mention the 38 sig fig limit in 128-bit decimal documentation above (and if I'm calculating this correctly that the precision limit for 256-bit decimals is 76)?

File: cpp/src/arrow/type.h [hunk 2]
```diff
@@ -999,6 +1031,20 @@ class ARROW_EXPORT SparseUnionType : public UnionType {
   std::string name() const override { return "sparse_union"; }
 };
 
+/// \brief Concrete type class for sparse union data
```
Comment (jonkeane) [inline cpp/src/arrow/type.h:1034]:
```suggestion
/// \brief Concrete type class for dense union data
```

?

File: python/pyarrow/types.pxi [hunk 3]
```diff
@@ -2110,11 +2110,28 @@ def float64():
...
+    decimal type can represent; the scale is the number of digits after
+    the decimal point (note the scale can be negative).
+
+    As an example, ``decimal128(7, 3)`` can exactly represent the numbers
+    1234.567 and -1234.567 (encoded internally as the 128-bit integers
+    1234567 and -1234567, respectively), but neither 12345.67 nor 123.4567.
+
+    ``decimal128(5, -3)`` can exactly represent the number 12345000
+    (encoded internally as the 128-bit integer 12345), but neither
+    123456000 nor 12345600.
```
Comment (jonkeane) [inline python/pyarrow/types.pxi:2138]: It also would be unable to represent the numbers 123450000 and 1234500 correct? Would it be clearer to use those similar to the decimal example we use above in type.h? Where the sig figs are exactly the same, but the decimal is the only thing that is different?

File: python/pyarrow/types.pxi [hunk 4]
```diff
@@ -2386,25 +2426,102 @@ def struct(fields):
     return pyarrow_wrap_data_type(struct_type)
 
 
+cdef _extract_union_params(children_fields, type_codes,
```
Comment (jonkeane) [inline python/pyarrow/types.pxi:2429]: This is super nit-picky, but should the first arg be `child_fields`? That would match what's in the body when iterating over them `child_fields` yields a single `child_field` (as opposed to yielding a `children_field`).

File: python/pyarrow/types.pxi [hunk 5]
```diff
@@ -2386,25 +2426,102 @@ def struct(fields):
...
+def sparse_union(children_fields, type_codes=None):
+    """
+    Create SparseUnionType from children fields.
+
+    A union is defined by an ordered sequence of child types; each slot in
+    the union can have a value chosen from these types.
+
+    Parameters
+    ----------
+    fields : sequence of Field values
```
Comment (jonkeane) [inline python/pyarrow/types.pxi:2457]: This should be `children_fields : ...` to match the signature, yeah? (or `child_fields` if we go that naming convention)

**Review by jonkeane — APPROVED** [review]
A few suggestions, but LGTM

File: python/pyarrow/types.pxi (hunk 5 above)
Comment (pitrou) [reply python/pyarrow/types.pxi:2457]: What is better? `children_fields` or `child_fields`?

Comment (jonkeane) [reply python/pyarrow/types.pxi:2457]: `child_fields` sounds better to me and matches what's inside the function

**Comment by pitrou** [comment]
Thanks @jonkeane . I've now applied your suggestions.

**Comment by ARF1** [comment]
@pitrou Many thanks for the explanation of the `scale` and `precision` parameters. The new python docstring is very clear.

File: python/pyarrow/types.pxi [hunk 6]
```diff
@@ -2386,25 +2426,102 @@ def struct(fields):
...
+
+def sparse_union(child_fields, type_codes=None):
     """
-    Create UnionType from children fields.
+    Create SparseUnionType from child fields.
 
-    A union is defined by an ordered sequence of types; each slot in the union
-    can have a value chosen from these types.
+    A union is defined by an ordered sequence of child types; each slot in
+    the union can have a value chosen from these types.
```
Comment (jorisvandenbossche) [inline python/pyarrow/types.pxi:2454]: Do we want to add the expanded explanation about sparse (or dense) unions, that you added to the C++ code? (don't know to what extent we want to duplicate this content, but right now the python docs don't have much explanation about it)

**Review by jorisvandenbossche — APPROVED** [review]
Nice docstrings!

File: python/pyarrow/types.pxi (hunk 6 above)
Comment (pitrou) [reply python/pyarrow/types.pxi:2454]: I can and replace this short description with it.

**Comment by pitrou** [comment]
I'll merge if CI is green. Thanks for the fast reviews!

**Pull request closed without merging**